packages/modal/src/positioning/ux-modal-positioning.ts

```diff
@@ -0,0 +1,409 @@
+import { UxPositioningOptions, UxModalPlacement, UxModalMissingSpaceStrategy } from './interfaces';
+import { inject, TaskQueue } from 'aurelia-framework';
+// import this CSS for the default hidden class `.ux-positioning--hidden`
+import './ux-modal-positioning.css';
+import { getLogger } from 'aurelia-logging';
+const log = getLogger('ux-modal-positioning');
+
+type MainPlacement = 'left' | 'right' | 'bottom' | 'top';
+type SecondaryPlacement = 'vstart' | 'vcenter' | 'vend' | 'hstart' | 'hcenter' | 'hend';
+const flip: {[key: string]: MainPlacement} = {
+  'left': 'right',
+  'right': 'left',
+  'bottom': 'top',
+  'top': 'bottom'
+};
+
+@inject(TaskQueue)
+export class UxModalPositioning {
+
+  public anchor: HTMLElement;
+  public element: HTMLElement;
+  public constraintElement: HTMLElement | Window;
+  public preferedPlacement: UxModalPlacement = 'auto';
+  public missingSpaceStrategy: UxModalMissingSpaceStrategy = 'flip';
+  public offsetX: number = 5;
+  public offsetY: number = 5;
+  public constraintMarginX: number = 5;
+  public constraintMarginY: number = 5;
+  public hiddenClass = 'ux-positioning--hidden';
+
+
+  // will be used for each positioning calculation
+  private anchorRect: DOMRect;
+  private elementRect: DOMRect;
+  private hostOffsetX: number = 0;
+  private hostOffsetY: number = 0;
+  private constraintX: number = 0;
+  private constraintY: number = 0;
+  private constraintHeight: number = 0;
+  private constraintWidth: number = window.innerWidth;
+  private spaceTop?: number = void 0;
+  private spaceRight?: number = void 0;
+  private spaceBottom?: number = void 0;
+  private spaceLeft?: number = void 0;
+
+  constructor(public taskQueue: TaskQueue) {}
+
+  public getInstance(anchor: HTMLElement, element: HTMLElement, options?: UxPositioningOptions) {
```

Comment:
Should we make this method `static`, with an extra parameter for `TaskQueue` instead?

Reply:
My question (3) about factoring was linked to your question here. I would prefer to have a static `getInstance` or `create` method. In order to go in this direction I suggest to drop the TaskQueue dependency and replace the only usage by a `setTimeout(/*..*/, 0)` which has the same effect (tested).

This would leave us with a service that has no dependency and that can be used in any application (aurelia or not)

What do you think  @bigopon  ?

Reply:
Thanks, I see what you meant with 3 now. I was vaguely guessing that as well. For that, I think we can use an interface here instead. Instead of expecting a `TaskQueue`, we can expect an object with following shape:
```ts
interface ITaskQueue {
  queueTask(task: () => void): void;
}
```
maybe that'd work as a first step? We don't have to solve all the reuse-ability issues for now.

Reply:
Also, folks still have Popper if they want something.

Reply:
If we go static I don't think it's a good idea to expect anything like a TaskQueue. As a consumer of this service I would not want to always have to pass on a taskqueue instance. This should be handled by the service.

We could however add an option in the `options` parameter where one could pass on the taskQueue if wanted, otherwise it defaults to `setTimeout(..., 0)`.

Or we can keep it non-static and have a stronger dependency with Aurelia codebase.

Also, the taskQueue is in fact only needed for a very specific use case which do not occur all the time and only happens once during the init phase. So placing here a timeout 0 doesn't hurt much ?!?

Reply:
Ok lets go with a timeout for now and add TQ later. The issue I have with it is its current way of init: create a empty instance, and then passing the injected task queue around. It feels patchy/hacky to me. Maybe lets ping @EisenbergEffect if he has any input around the API

Reply:
Yes, it is patchy/hacky and that was where I was looking for a good factory pattern. If we go static then we don't  need anything more. Otherwise we might need two classes, first a factory service that create a proper instance of the actual positioning class.

Reply:
If you use the factory resolver, it can give you a function that will create an instance of UxModalPositioning, allowing you to supply the anchor, element, and options.

See https://github.com/aurelia/dependency-injection/blob/master/src/resolvers.ts#L341

Reply:
Thanks @EisenbergEffect for pointing to the `Factory` resolver.